jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/OnDiskGraphIndexWriter.java

```diff
@@ -0,0 +1,210 @@
+            for (int i = 0; i < oldToNewOrdinals.size(); i++) {
+                var entry = entriesByNewOrdinal.get(i);
+                int originalOrdinal = entry.getKey();
+                int newOrdinal = entry.getValue();
+                if (!graph.containsNode(originalOrdinal)) {
+                    continue;
+                }
+
+                out.writeInt(newOrdinal); // unnecessary, but a reasonable sanity check
+
+                for (var feature : featureWriters.values()) {
+                    feature.writeInline(originalOrdinal, out);
+                }
+
+                var neighbors = view.getNeighborsIterator(originalOrdinal);
+                out.writeInt(neighbors.size());
+                int n = 0;
+                for (; n < neighbors.size(); n++) {
+                    out.writeInt(oldToNewOrdinals.get(neighbors.nextInt()));
+                }
+                assert !neighbors.hasNext();
+
+                // pad out to maxEdgesPerNode
+                for (; n < graph.maxDegree(); n++) {
+                    out.writeInt(-1);
+                }
+            };
+        } catch (Exception e) {
+            throw new IOException(e);
+        }
+    }
+
+    /**
+     * @return a Map of old to new graph ordinals where the new ordinals are sequential starting at 0,
+     * while preserving the original relative ordering in `graph`.  That is, for all node ids i and j,
+     * if i &lt; j in `graph` then map[i] &lt; map[j] in the returned map.
+     */
+    public static Map<Integer, Integer> getSequentialRenumbering(GraphIndex graph) {
+        try (var view = graph.getView()) {
+            Int2IntHashMap oldToNewMap = new Int2IntHashMap(-1);
+            int nextOrdinal = 0;
+            for (int i = 0; i < view.getIdUpperBound(); i++) {
+                if (graph.containsNode(i)) {
+                    oldToNewMap.put(i, nextOrdinal++);
+                }
+            }
+            return oldToNewMap;
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Builder for OnDiskGraphIndexWriter, with optional features.
+     */
+    public static class Builder {
+        private GraphIndex graphIndex;
+        private Map<Integer, Integer> oldToNewOrdinals;
+        private PQVectors pqVectors = null;
+        private RandomAccessVectorValues ravv = null;
+        private LocallyAdaptiveVectorQuantization lvq = null;
+        private LocallyAdaptiveVectorQuantization.QuantizedVector[] lvqVectors = null;
+
+        public Builder(GraphIndex graphIndex) {
+            this(graphIndex, getSequentialRenumbering(graphIndex));
+        }
+
+        public Builder(GraphIndex graphIndex, Map<Integer, Integer> oldToNewOrdinals) {
+            this.graphIndex = graphIndex;
+            this.oldToNewOrdinals = oldToNewOrdinals;
+        }
+
+        public Builder withFusedADC(PQVectors pqVectors) {
```

Comment:
WDYT about having instead a with(Feature) method so we don't have to hardcode different pieces of state in the Builder and Writer?